File: packages/i18nify-js/src/modules/names/__tests__/isValidName.test.ts
import { isValidName } from '../index';

describe('names - isValidName', () => {
  describe('valid names', () => {
    it('returns valid for a normal full name', () => {
      expect(isValidName('Ada Lovelace')).toEqual({ isValid: true });
    });

    it('returns valid for names with non-ASCII letters', () => {
      expect(isValidName('José Álvarez')).toEqual({ isValid: true });
      expect(isValidName('निखिल')).toEqual({ isValid: true });
    });

    it('trims surrounding whitespace', () => {
      expect(isValidName('  Grace Hopper  ')).toEqual({ isValid: true });
    });
  });

  describe('invalid names', () => {
    it('rejects names shorter than validation_rules.min_length', () => {
      expect(isValidName('A')).toEqual({
        isValid: false,
        reason: 'too_short',
      });
    });

    it('rejects names longer than validation_rules.max_length', () => {
      expect(isValidName('A'.repeat(101))).toEqual({
        isValid: false,
        reason: 'too_long',
      });
    });

    it('rejects blocklisted placeholder names', () => {
      expect(isValidName('test')).toEqual({
        isValid: false,
        reason: 'blocklisted',
      });
    });

    it('rejects sequential character runs', () => {
      expect(isValidName('pqrs')).toEqual({
        isValid: false,
        reason: 'sequential_chars',
      });
    });

    it('rejects sequential character runs split by spaces', () => {
      expect(isValidName('a b c d')).toEqual({
        isValid: false,
        reason: 'sequential_chars',
      });
    });

    it('rejects repeated character runs', () => {
      expect(isValidName('Sss Sharma')).toEqual({
        isValid: false,
        reason: 'repeating_chars',
      });
    });

    it('rejects values where letters are not dominant', () => {
      expect(isValidName('12@#$ab')).toEqual({
        isValid: false,
        reason: 'non_alpha_dominant',
      });
    });
  });

  describe('options', () => {
    it('uses a custom blocklist as a replacement by default', () => {
      expect(isValidName('test', { blocklist: ['blocked'] })).toEqual({
        isValid: true,
      });
      expect(isValidName('blocked', { blocklist: ['blocked'] })).toEqual({
        isValid: false,
        reason: 'blocklisted',
      });
    });

    it('can extend the default blocklist', () => {
      expect(
        isValidName('blocked', {
          blocklist: ['blocked'],
          allowBlocklistExtension: true,
        }),
      ).toEqual({
        isValid: false,
        reason: 'blocklisted',
      });
      expect(
        isValidName('test', {
          blocklist: ['blocked'],
          allowBlocklistExtension: true,
        }),
      ).toEqual({
        isValid: false,
        reason: 'blocklisted',
      });
    });

    it('uses custom sequential and repeating thresholds', () => {
      expect(isValidName('wxy', { sequentialThreshold: 3 })).toEqual({
        isValid: false,
        reason: 'sequential_chars',
      });
      expect(isValidName('mnop', { sequentialThreshold: 5 })).toEqual({
        isValid: true,
      });
      expect(isValidName('Annaa', { repeatingThreshold: 2 })).toEqual({
        isValid: false,
        reason: 'repeating_chars',
      });
    });

    it('uses a custom alpha dominance threshold', () => {
      expect(
        isValidName('a1 b2', {
          alphaDominanceThreshold: 0.75,
        }),
      ).toEqual({
        isValid: false,
        reason: 'non_alpha_dominant',
      });
    });
  });

  describe('invalid arguments', () => {
    it('throws for an empty string', () => {
      expect(() => isValidName('')).toThrow(/name/);
    });

    it('throws for a whitespace-only string', () => {
      expect(() => isValidName('   ')).toThrow(/name/);
    });

    it('throws for a non-string argument', () => {
      expect(() => isValidName(123 as any)).toThrow(/name/);
    });
  });
});

File: packages/i18nify-go/modules/names/is_valid_name_test.go
package names

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestIsValidName(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		opts    *IsValidNameOptions
		want    NameValidationResult
		wantErr bool
	}{
		{
			name:  "normal full name",
			input: "Ada Lovelace",
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "unicode letters",
			input: "निखिल",
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "accented latin letters",
			input: "José Álvarez",
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "surrounding whitespace is trimmed",
			input: "  Grace Hopper  ",
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "blocklisted placeholder",
			input: "test",
			want:  NameValidationResult{IsValid: false, Reason: "blocklisted"},
		},
		{
			name:  "shorter than validation rules minimum",
			input: "A",
			want:  NameValidationResult{IsValid: false, Reason: "too_short"},
		},
		{
			name:  "longer than validation rules maximum",
			input: strings.Repeat("A", 101),
			want:  NameValidationResult{IsValid: false, Reason: "too_long"},
		},
		{
			name:  "sequential characters",
			input: "pqrs",
			want:  NameValidationResult{IsValid: false, Reason: "sequential_chars"},
		},
		{
			name:  "sequential characters split by spaces",
			input: "a b c d",
			want:  NameValidationResult{IsValid: false, Reason: "sequential_chars"},
		},
		{
			name:  "repeating characters",
			input: "Sss Sharma",
			want:  NameValidationResult{IsValid: false, Reason: "repeating_chars"},
		},
		{
			name:  "letters are not dominant",
			input: "12@#$ab",
			want:  NameValidationResult{IsValid: false, Reason: "non_alpha_dominant"},
		},
		{
			name:  "custom blocklist replaces defaults",
			input: "test",
			opts:  &IsValidNameOptions{Blocklist: []string{"blocked"}},
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "custom blocklist value",
			input: "blocked",
			opts:  &IsValidNameOptions{Blocklist: []string{"blocked"}},
			want:  NameValidationResult{IsValid: false, Reason: "blocklisted"},
		},
		{
			name:  "custom blocklist extends defaults",
			input: "test",
			opts: &IsValidNameOptions{
				Blocklist:               []string{"blocked"},
				AllowBlocklistExtension: true,
			},
			want: NameValidationResult{IsValid: false, Reason: "blocklisted"},
		},
		{
			name:  "extended custom blocklist value",
			input: "blocked",
			opts: &IsValidNameOptions{
				Blocklist:               []string{"blocked"},
				AllowBlocklistExtension: true,
			},
			want: NameValidationResult{IsValid: false, Reason: "blocklisted"},
		},
		{
			name:  "custom sequential threshold",
			input: "wxy",
			opts:  &IsValidNameOptions{SequentialThreshold: 3},
			want:  NameValidationResult{IsValid: false, Reason: "sequential_chars"},
		},
		{
			name:  "custom sequential threshold can relax default",
			input: "mnop",
			opts:  &IsValidNameOptions{SequentialThreshold: 5},
			want:  NameValidationResult{IsValid: true},
		},
		{
			name:  "custom repeating threshold",
			input: "Annaa",
			opts:  &IsValidNameOptions{RepeatingThreshold: 2},
			want:  NameValidationResult{IsValid: false, Reason: "repeating_chars"},
		},
		{
			name:  "custom alpha dominance threshold",
			input: "a1 b2",
			opts:  &IsValidNameOptions{AlphaDominanceThreshold: 0.75},
			want:  NameValidationResult{IsValid: false, Reason: "non_alpha_dominant"},
		},
		{
			name:    "empty string returns error",
			input:   "",
			wantErr: true,
		},
		{
			name:    "whitespace-only returns error",
			input:   "   ",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := IsValidName(tt.input, tt.opts)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}

			require.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}
